src/librustdoc/html/static/js/search.js

```diff
@@ -76,39 +76,105 @@ function printTab(nb) {
-    if (s1 === s2) {
-        return 0;
+
+/*
+ * This function was translated, mostly line-for-line, from
+ * https://github.com/rust-lang/rust/blob/ff4b772f805ec1e/compiler/rustc_span/src/edit_distance.rs
+ *
+ * The current implementation is the restricted Damerau-Levenshtein algorithm. It is restricted
+ * because it does not permit modifying characters that have already been transposed. The specific
+ * algorithm should not matter to the caller of the methods, which is why it is not noted in the
+ * documentation.
+ */
+let editDistanceCurrent = [];
+let editDistancePrev = [];
+let editDistancePrevPrev = [];
+function editDistance(a, b, limit) {
+    // Ensure that `b` is the shorter string, minimizing memory use.
+    if (a.length < b.length) {
+        const aTmp = a;
+        a = b;
+        b = aTmp;
+    }
+
+    const minDist = a.length - b.length;
+    // If we know the limit will be exceeded, we can return early.
+    if (minDist > limit) {
+        return limit + 1;
+    }
+
+    // Strip common prefix.
+    // We know that `b` is the shorter string, so we don't need to check
+    // `a.length`.
+    while (b.length > 0 && b[0] === a[0]) {
+        a = a.substring(1);
+        b = b.substring(1);
+    }
+    // Strip common suffix.
+    while (b.length > 0 && b[b.length - 1] === a[a.length - 1]) {
+        a = a.substring(0, a.length - 1);
+        b = b.substring(0, b.length - 1);
     }
-    const s1_len = s1.length, s2_len = s2.length;
-    if (s1_len && s2_len) {
-        let i1 = 0, i2 = 0, a, b, c, c2;
-        const row = levenshtein_row2;
-        while (i1 < s1_len) {
-            row[i1] = ++i1;
-        }
-        while (i2 < s2_len) {
-            c2 = s2.charCodeAt(i2);
-            a = i2;
-            ++i2;
-            b = i2;
-            for (i1 = 0; i1 < s1_len; ++i1) {
-                c = a + (s1.charCodeAt(i1) !== c2 ? 1 : 0);
-                a = row[i1];
-                b = b < a ? (b < c ? b + 1 : c) : (a < c ? a + 1 : c);
-                row[i1] = b;
-            }
-        }
-        return b;
+
+    // If either string is empty, the distance is the length of the other.
+    // We know that `b` is the shorter string, so we don't need to check `a`.
+    if (b.length === 0) {
+        return minDist;
+    }
+
+    const aLength = a.length;
+    const bLength = b.length;
+
+    for (let i = 0; i <= bLength; ++i) {
+        editDistanceCurrent[i] = 0;
+        editDistancePrev[i] = i;
+        editDistancePrevPrev[i] = Number.MAX_VALUE;
     }
-    return s1_len + s2_len;
+
+    // row by row
+    for (let i = 1; i <= aLength; ++i) {
+        editDistanceCurrent[0] = i;
+        const aIdx = i - 1;
+
+        // column by column
+        for (let j = 1; j <= bLength; ++j) {
+            const bIdx = j - 1;
+
+            // There is no cost to substitute a character with itself.
+            const substitutionCost = a[aIdx] === b[bIdx] ? 0 : 1;
+
+            editDistanceCurrent[j] = Math.min(
```

Comment:
You never `push` to `editDistanceCurrent`, so I don't understand how this can work. Same for `editDistancePrev` and `editDistancePrevPrev`.

Reply:
Wow. TIL.